crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,234 @@
+            //    C ░░░░░░░░░▒▒▒▒▒▒▒
+            //               |t0 + (duration - clbit_write_latency)
+
+            let t1 = t0 + op_duration;
+            for clbit in cargs.iter() {
+                idle_before.insert(*clbit, t0 + (op_duration - clbit_write_latency as f64));
+            }
+            t1
+        } else {
+            // Directives (like Barrier)
+            let t0 = qargs
+                .iter()
+                .chain(cargs.iter())
+                .map(|bit| *idle_before.get(bit).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+            t0 + op_duration
+        };
+
+        for qubit in qargs {
+            idle_before.insert(qubit, t1);
+        }
+
+        node_start_time.insert(node_index, t1);
+    }
+
+    // Compute maximum instruction available time
+    let circuit_duration = *idle_before
+        .values()
+        .max_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap_or(&0.0);
+
+    // Note that ALAP pass is inversely scheduled, thus
+    // t0 is computed by subtracting t1 from the entire circuit duration.
+
+    let mut result: HashMap<NodeIndex, TimeValue> = HashMap::new();
+    for (node_idx, t1) in node_start_time {
+        let final_time = match node_durations.get(&node_idx) {
+            Some(TimeValue::Int(_)) => TimeValue::from((circuit_duration - t1).round() as u64),
+            Some(TimeValue::Float(_)) => TimeValue::from(circuit_duration - t1),
+            None => return Err(TranspilerError::new_err("No duration for node")),
+        };
+        result.insert(node_idx, final_time);
+    }
+
+    Ok(result)
+}
+
+#[pyfunction]
+/// Runs the ALAPSchedule analysis pass on dag.
+///
+/// Args:
+///     dag (DAGCircuit): DAG to schedule.
+///     clbit_write_latency (u64): The latency to write classical bits.
+///     node_durations (HashMap<usize, f64>): Mapping from node indices to operation durations.
+///
+/// Returns:
+///     PyDict: A dictionary mapping each DAGOpNode to its scheduled start time.
+///
+#[pyo3(name = "alap_schedule_analysis", signature= (dag, clbit_write_latency, node_durations))]
+pub fn py_run_alap_schedule_analysis(
+    py: Python,
+    dag: &DAGCircuit,
+    clbit_write_latency: u64,
+    node_durations: &Bound<PyDict>,
+) -> PyResult<Py<PyDict>> {
+    // Extract indices and durations from PyDict
+    let mut op_durations: HashMap<NodeIndex, TimeValue> = HashMap::new();
+
+    for (py_node, py_duration) in node_durations.iter() {
+        let node_idx = py_node
+            .downcast_into::<DAGOpNode>()?
+            .extract::<DAGNode>()?
+            .node
+            .expect("Node index not found.");
+
+        let op_duration = if let Ok(val) = py_duration.extract::<u64>() {
+            val.into()
+        } else if let Ok(val) = py_duration.extract::<f64>() {
+            val.into()
+        } else {
+            return Err(TranspilerError::new_err("Duration must be u64 or f64"));
```

Comment:
```suggestion
            return Err(TranspilerError::new_err("Duration must be int or float"));
```
Nitpick: Since this error message is only exposed to Python, we can use `int` or `float` instead of the rust types.